libs/hooks/debounce/src/lib/debounce-hook.ts

```diff
@@ -0,0 +1,190 @@
+   */
+  debounceTime: number;
+  /**
+   * Max number of items to be kept in cache before the oldest entry falls out.
+   */
+  maxCacheItems: number;
+};
+
+/**
+ * A hook that wraps another hook and debounces its execution based on the provided options.
+ * Each stage of the hook (before, after, error, finally) is debounced independently.
+ * If a stage is called with a cache key that has been seen within the debounce time, the inner hook's stage will not run.
+ * If no cache key supplier is provided for a stage, that stage will always run.
+ */
+export class DebounceHook<T extends FlagValue = FlagValue> implements Hook {
+  private readonly cache: FixedSizeExpiringCache<true | CachedError>;
+  private readonly cacheErrors: boolean;
+
+  public constructor(
+    private readonly innerHook: Hook,
+    private readonly options: Options<T>,
+  ) {
+    this.cacheErrors = options.cacheErrors || false;
+    this.cache = new FixedSizeExpiringCache<true | CachedError>({
+      maxItems: options.maxCacheItems,
+      ttlMs: options.debounceTime,
+    });
+  }
+
+  before(hookContext: HookContext, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'before',
+      () => this.options?.beforeCacheKeySupplier?.(hookContext.flagKey, hookContext.context),
+      () => this.innerHook?.before?.(hookContext, hookHints),
+    );
+  }
+
+  after(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'after',
+      () => this.options?.afterCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.after?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  error(hookContext: HookContext, err: unknown, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'error',
+      () => this.options?.errorCacheKeySupplier?.(hookContext.flagKey, hookContext.context, err),
+      () => this.innerHook?.error?.(hookContext, err, hookHints),
+    );
+  }
+
+  finally(hookContext: HookContext, evaluationDetails: EvaluationDetails<T>, hookHints?: HookHints) {
+    this.maybeSkipAndCache(
+      'finally',
+      () => this.options?.finallyCacheKeySupplier?.(hookContext.flagKey, hookContext.context, evaluationDetails),
+      () => this.innerHook?.finally?.(hookContext, evaluationDetails, hookHints),
+    );
+  }
+
+  private maybeSkipAndCache(
+    stage: 'before' | 'after' | 'error' | 'finally',
+    keyGenCallback: () => string | null | undefined,
+    hookCallback: () => void,
+  ) {
+    // the cache key is a concatenation of the result of calling keyGenCallback and the stage
+    const dynamicKey = keyGenCallback();
+
+    // if the keyGenCallback returns nothing, we don't do any caching
+    if (dynamicKey) {
+      const cacheKeySuffix = stage;
+      const cacheKey = `${dynamicKey}::${cacheKeySuffix}`;
```

Comment:
I guess the colons aren't necessary with the current list of stages. However, it's probably worth keeping it still.